File: Human/Human.cpp
#include "Human.h"
#include <stdexcept>
#include "../Settings.h"

Human::Human(World *world, int x, int y, int i, int i1) :
    Animal(HUMAN_STRENGTH,
           HUMAN_INITIATIVE,
           HUMAN_SYMBOL,
           HUMAN_COLOR,
           world,
           x,
           y,
           -1,
           -1) { }

void Human::action() {
  this->setPreviousX(this->getPosition().getX());
  this->setPreviousY(this->getPosition().getY());

  switch (this->getDirection()) {
    case 'u':
      this->setPosition(this->getPosition().getX(), this->getPosition().getY() - 1);
      break;
    case 'd':
      this->setPosition(this->getPosition().getX(), this->getPosition().getY() + 1);
      break;
    case 'r':
      this->setPosition(this->getPosition().getX() + 1, this->getPosition().getY());
      break;
    case 'l':
      this->setPosition(this->getPosition().getX() - 1, this->getPosition().getY());
      break;
    default:
      return;
  }

  Organism* organism = this->getWorld()->getOrganismAtPosition(this, this->getPosition().getX(), this->getPosition().getY());
  if (organism != nullptr && organism->getStrength() > this->getStrength() && this->getAbilityDuration() > 0) {
    Position newPosition = this->getWorld()->generateNewSafePosition(this, this->getPosition().getX(), this->getPosition().getY());
    this->getWorld()->addLog("%s POS:(%d, %d) moved to (%d, %d)",
                             this->getType().c_str(),
                             this->getPreviousX(),
                             this->getPreviousY(),
                             this->getPosition().getX(),
                             this->getPosition().getY());

    this->setPosition(newPosition.getX(), newPosition.getY());

    this->getWorld()->addLog("%s POS:(%d, %d) run away to (%d, %d)",
                             this->getType().c_str(),
                             this->getPreviousX(),
                             this->getPreviousY(),
                             this->getPosition().getX(),
                             this->getPosition().getY());
    return;
  }

  this->getWorld()->addLog("%s POS:(%d, %d) moved to (%d, %d)",
                           this->getType().c_str(),
                           this->getPreviousX(),
                           this->getPreviousY(),
                           this->getPosition().getX(),
                           this->getPosition().getY());
}

void Human::collision(Organism *organism) {
  Animal::collision(organism);
}

Organism* Human::createChild(int x, int y, int parent1Id, int parent2Id) {
  throw std::runtime_error("Human cannot reproduce");
}

std::string Human::getType() const { return "Human"; }

bool Human::deflectedAttack(Organism *attacker) {
  if (this->getAbilityDuration() > 0) return true;
  return Animal::deflectedAttack(attacker);
}

char Human::getDirection() const { return this->direction; }
void Human::setDirection(char d) { this->direction = d; }

int Human::getAbilityDuration() const { return this->abilityDuration; }
int Human::getAbilityCooldown() const { return this->abilityCooldown; }

int Human::getDefaultAbilityDuration() const { return this->defaultAbilityDuration; }
int Human::getDefaultAbilityCooldown() const { return this->defaultAbilityCooldown; }

void Human::setAbilityDuration(int duration) { this->abilityDuration = duration; }
void Human::setAbilityCooldown(int cooldown) { this->abilityCooldown = cooldown; }

File: src/Human/Human.java
package Human;

import Animal.Animal;
import Field.Field;
import Organism.Organism;
import Organism.OrganismData;
import Settings.Settings;
import World.World;

public class Human extends Animal {

    private int abilityDuration;
    private int abilityCooldown;
    private MoveDirection moveDirection;

    public Human(World world, Field field, int parent1Id, int parent2Id) {
        super(new OrganismData(world.getNextOrganismId(), Settings.HUMAN_INITIATIVE, Settings.HUMAN_STRENGTH, Settings.HUMAN_IMAGE, parent1Id, parent2Id), world, field);
    }

    @Override
    public void action() {
        this.lastField = this.currentField;
        this.currentField.setOrganism(null);

        switch (this.getMoveDirection()) {
            case MoveDirection.Up:
                this.currentField = this.getWorld().getField(this.currentField.getNumber() - 1);
                break;
            case MoveDirection.Down:
                this.currentField = this.getWorld().getField(this.currentField.getNumber() + 1);
                break;
            case MoveDirection.Left:
                this.currentField = this.getWorld().getField(this.currentField.getNumber() - this.getWorld().getWorldSettings().height());
                break;
            case MoveDirection.Right:
                this.currentField = this.getWorld().getField(this.currentField.getNumber() + this.getWorld().getWorldSettings().height());
                break;
            case MoveDirection.UP_LEFT:
                this.currentField = this.getWorld().getField(this.currentField.getNumber() - 1);
                break;
            case MoveDirection.UP_RIGHT:
                this.currentField = this.getWorld().getField(this.currentField.getNumber() + this.getWorld().getWorldSettings().height() - 1);
                break;
            case MoveDirection.DOWN_LEFT:
                this.currentField = this.getWorld().getField(this.currentField.getNumber() - this.getWorld().getWorldSettings().height() + 1);
                break;
            case MoveDirection.DOWN_RIGHT:
                this.currentField = this.getWorld().getField(this.currentField.getNumber() + 1);
                break;
            default:
                break;
        }

        this.currentField.setOrganism(this);
    }

    @Override
    public void reproduce(Field field, int parent1Id, int parent2Id) {
        throw new UnsupportedOperationException("Human cannot reproduce.");
    }

    @Override
    public String getType() {
        return "Human";
    }

    @Override
    public boolean deflectedAttack(Organism attacker) {
        if (this.getAbilityDuration() > 0) return true;
        return super.deflectedAttack(attacker);
    }

    public int getAbilityDuration() {
        return this.abilityDuration;
    }

    public void setAbilityDuration(int abilityDuration) {
        this.abilityDuration = abilityDuration;
    }

    public int getAbilityCooldown() {
        return this.abilityCooldown;
    }

    public void setAbilityCooldown(int abilityCooldown) {
        this.abilityCooldown = abilityCooldown;
    }

    public MoveDirection getMoveDirection() {
        return this.moveDirection;
    }

    public void setMoveDirection(MoveDirection moveDirection) {
        this.moveDirection = moveDirection;
    }


}
